src/runtime-rs/crates/hypervisor/src/dragonball/vmm_instance.rs

```diff
@@ -0,0 +1,350 @@
+    }
+    pub fn pid(&self) -> u32 {
+        std::process::id()
+    }
+
+    pub fn stop(&mut self) -> Result<()> {
+        self.handle_request(Request::Sync(VmmAction::ShutdownMicroVm))
+            .map_err(|e| {
+                warn!(sl!(), "Failed to shutdown MicroVM. {}", e);
+                e
+            })
+            .ok();
+        // vmm is not running, join thread will be hang.
+        if self.is_uninitialized() || self.vmm_thread.is_none() {
+            debug!(sl!(), "vmm-master thread is uninitialized or has exited.");
+            return Ok(());
+        }
+        debug!(sl!(), "join vmm-master thread exit.");
+
+        // vmm_thread must be exited, otherwise there will be other sync issues.
+        // unwrap is safe, if vmm_thread is None, impossible run to here.
+        self.vmm_thread.take().unwrap().join().ok();
+        info!(sl!(), "vmm-master thread join succeed.");
+        Ok(())
+    }
+
+    fn send_request(&self, vmm_action: VmmAction) -> Result<VmmResponse> {
+        if let Some(ref to_vmm) = self.to_vmm {
+            to_vmm
+                .send(Box::new(vmm_action.clone()))
+                .with_context(|| format!("Failed to send  {:?} via channel ", vmm_action))?;
+        } else {
+            return Err(anyhow!("to_vmm is None"));
+        }
+
+        //notify vmm action
+        if let Err(e) = self.to_vmm_fd.write(1) {
+            return Err(anyhow!("failed to notify vmm. {}", e));
+        }
+
+        if let Some(from_vmm) = self.from_vmm.as_ref() {
+            match from_vmm.recv() {
+                Err(e) => Err(anyhow!("vmm recv err. {}", e)),
+                Ok(vmm_outcome) => Ok(vmm_outcome),
+            }
+        } else {
+            Err(anyhow!("from_vmm is None"))
+        }
+    }
+
+    fn handle_request(&self, req: Request) -> Result<VmmData> {
+        let Request::Sync(vmm_action) = req;
+        match self.send_request(vmm_action) {
+            Ok(vmm_outcome) => match *vmm_outcome {
+                Ok(vmm_data) => Ok(vmm_data),
+                Err(vmm_action_error) => Err(anyhow!("vmm action error: {:?}", vmm_action_error)),
+            },
+            Err(e) => Err(e),
+        }
+    }
+
+    fn handle_request_with_retry(&self, req: Request) -> Result<VmmData> {
+        let Request::Sync(vmm_action) = req;
+        for count in 0..REQUEST_RETRY {
+            match self.send_request(vmm_action.clone()) {
+                Ok(vmm_outcome) => match *vmm_outcome {
+                    Ok(vmm_data) => {
+                        info!(
+                            sl!(),
+                            "success to send {:?} after retry {}", &vmm_action, count
+                        );
+                        return Ok(vmm_data);
+                    }
+                    Err(vmm_action_error) => {
+                        if let VmmActionError::UpcallNotReady = vmm_action_error {
+                            std::thread::sleep(std::time::Duration::from_millis(10));
```

Comment:
Question: The retry interval (10ms)  here and above `wait_vmm_ready` is big or small, or just a fair suitable value?

Or can it be a configurable value?

I just want to open a discussion, you need not fix/modify it in this PR.

Reply:
Thanks for your review, rise an issue #4635